src/librustc_borrowck/borrowck/mir/dataflow/mod.rs

```diff
@@ -0,0 +1,490 @@
+        let offset = self.words_per_block * block_idx;
+        let range = E::new(offset)..E::new(offset + self.words_per_block);
+        BlockSets {
+            on_entry: self.on_entry_sets.bits.range_mut(&range),
+            gen_set: self.gen_sets.bits.range_mut(&range),
+            kill_set: self.kill_sets.bits.range_mut(&range),
+        }
+    }
+
+    fn lookup_set_for<'a>(&self, sets: &'a Bits<E>, block_idx: usize) -> &'a IdxSet<E> {
+        let offset = self.words_per_block * block_idx;
+        let range = E::new(offset)..E::new(offset + self.words_per_block);
+        sets.bits.range(&range)
+    }
+    pub fn gen_set_for(&self, block_idx: usize) -> &IdxSet<E> {
+        self.lookup_set_for(&self.gen_sets, block_idx)
+    }
+    pub fn kill_set_for(&self, block_idx: usize) -> &IdxSet<E> {
+        self.lookup_set_for(&self.kill_sets, block_idx)
+    }
+    pub fn on_entry_set_for(&self, block_idx: usize) -> &IdxSet<E> {
+        self.lookup_set_for(&self.on_entry_sets, block_idx)
+    }
+}
+
+/// Parameterization for the precise form of data flow that is used.
+pub trait DataflowOperator: BitwiseOperator {
+    /// Specifies the initial value for each bit in the `on_entry` set
+    fn bottom_value() -> bool;
+}
+
+pub trait BitDenotation {
+    /// Specifies what index type is used to access the bitvector.
+    type Idx: Idx;
+
+    /// Specifies what, if any, separate context needs to be supplied for methods below.
+    type Ctxt;
+
+    /// A name describing the dataflow analysis that this
+    /// BitDenotation is supporting.  The name should be something
+    /// suitable for plugging in as part of a filename e.g. avoid
+    /// space-characters or other things that tend to look bad on a
+    /// file system, like slashes or periods. It is also better for
+    /// the name to be reasonably short, again because it will be
+    /// plugged into a filename.
+    fn name() -> &'static str;
+
+    /// Size of each bitvector allocated for each block in the analysis.
+    fn bits_per_block(&self, &Self::Ctxt) -> usize;
+
+    /// Mutates the block-sets (the flow sets for the given
+    /// basic block) according to the effects that have been
+    /// established *prior* to entering the start block.
+    ///
+    /// (For example, establishing the call arguments.)
+    ///
+    /// (Typically this should only modify `sets.on_entry`, since the
+    /// gen and kill sets should reflect the effects of *executing*
+    /// the start block itself.)
+    fn start_block_effect(&self, ctxt: &Self::Ctxt, sets: &mut BlockSets<Self::Idx>);
+
+    /// Mutates the block-sets (the flow sets for the given
+    /// basic block) according to the effects of evaluating statement.
+    ///
+    /// This is used, in particular, for building up the
+    /// "transfer-function" represnting the overall-effect of the
+    /// block, represented via GEN and KILL sets.
+    ///
+    /// The statement here is `idx_stmt.1`; `idx_stmt.0` is just
```

Comment:
(note to self: this, like many comments, is out of date and needs to be fixed to reflect changes to the API here.)